### ISSUE TITLE
Unicode not supported in the self-built pcre

### ISSUE THREAD

**Issue by zxteloiv** [openresty/docker-openresty#22]
Specifically, this happens when I add the "u" option of a regular expression substitution.

``` lua
-- test.lua
newstr, n, err = ngx.re.gsub("测 试 用 例", "[ \t]", "", "ju")
ngx.say(newstr)

```

I use docker image tag `1.11.2.1-xenial`.

If the image is based on xenial / trusty, would it be better to install pcre and openssl from apt source, just like the openresty official document suggested?

``` bash
apt-get install libreadline-dev libncurses5-dev libpcre3-dev \
    libssl-dev perl make build-essential
```

**Comment by neomantra** [comment]
During the initial phases of this Docker initiative, we decided to build PCRE and OpenSSL from upstream sources.

I'm trying to understand the source of the problem, but am pretty unfamiliar with Unicode.  It seems that by default PCRE is [built with Unicode support](http://www.pcre.org/current/doc/html/pcre2build.html#SEC5) unless configured with  `--disable-unicode`.  But I don't see that directive anywhere in the OpenResty build process.

Is some other configuration needed?  Does your expression work with the `centos-rpm` variant?  Or does it work in Xenial if you build it yourself using `libpcre3-dev`?

**Comment by neomantra** [comment]
I just modified this test to print `err` as well and, with `xenial` and `alpine` variants, it emits:

`pcre_compile() failed: this version of PCRE is compiled without UTF support in "[ 	]" at "[ 	]"`

As noted, I'm pretty unfamiliar with Unicode.  Is this something that *should* actually be enabled with OpenResty?

**Comment by agentzh** [comment]
@neomantra Are you using the latest official APT repos for Ubuntu/Debian in the docker images?

http://openresty.org/en/linux-packages.html

We've enabled the unicode support in our official openresty-pcre package:

https://github.com/openresty/openresty-packaging/blob/master/deb/openresty-pcre/debian/rules#L31

**Comment by agentzh** [comment]
@neomantra The official OR RPM packages for PCRE should be the same.

**Comment by zxteloiv** [comment]
@neomantra I'm very sorry for the late reply because I haven't been working on that project for a while.

Despite the lack of official guidelines, it will be useful in practice to enable unicode in pcre when compiling from source.

For example, the dot `.` in a pattern will match a unicode character rather than a meanless byte, and `ngx.re.gmatch` will return an iterator which yields a complete unicode character (maybe 1~4 bytes in UTF-8) iteratively.
Meanwhile, not adding a `u` option to `ngx.re.match` will behave as usual even unicode is enabled.

**Comment by neomantra** [comment]
@agentzh I saw that those were released recently.  I intend to move to them (#19 is related) and was reviewing the `openresty-packaging` repo this morning.

**Comment by agentzh** [comment]
@neomantra Great. Thanks!

**Comment by neomantra** [comment]
With the `stretch` images released, the `stretch` and `centos-rpm` images have this resolved.

I still don't understand why it doesn't work in the other images.

**Comment by neomantra** [comment]
Fixed with fce29af4d81185eec00c7d68e1088e6b23a18d4f in tag `1.15.8.1-3`.
